### PR TITLE
Enhance the MVVM Implementation of the sign-in

### PR DESCRIPTION
**Pull request by aminebenji04**
In this PR I did a lot of refactors and feature adjustments to improve the structure and maintainability of the sign-in, specifically focusing on the checkUserProfile function and related components. 
###  Why I have done this?
- Improve separation of concerns in the sign-in code. 
- Moving checkUserProfile to the SignInViewModel as a private function makes it easier to use, and especially to test
- This new refactor will help us have easier mocking with hilt

### CODE REVIEW

File: app/src/test/java/com/android/sample/model/profile/MockProfilesRepository.kt [hunk 1]
```diff
@@ -0,0 +1,89 @@
+package com.android.sample.model.profile
+
+class MockProfilesRepository : ProfilesRepository {
+
```
Comment (GhaliBerbich) [inline app/src/test/java/com/android/sample/model/profile/MockProfilesRepository.kt:4]: Ok good refactoring

File: app/src/test/java/com/android/sample/model/profile/MockProfilesRepository.kt [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+  }
+
+  override fun addLikedActivity(
+      userId: String,
+      activityId: String,
+      onSuccess: () -> Unit,
+      onFailure: (Exception) -> Unit
+  ) {
+    TODO("Not yet implemented")
+  }
```
Comment (GhaliBerbich) [inline app/src/test/java/com/android/sample/model/profile/MockProfilesRepository.kt:55]: Is it someone else task to do this? If yes, maybe mention it on your PR description

File: app/src/test/java/com/android/sample/model/auth/SignInViewModelTests.kt [hunk 3]
```diff
@@ -222,14 +241,12 @@ class SignInViewModelTest {
...
-        // Advance coroutine
         advanceUntilIdle()
 
         // Then
-        verify(signInRepository)
-            .checkUserProfile(eq(uid), eq(navigationActions), eq(onAuthSuccess), eq(onAuthError))
+        verify(onProfileExists).invoke() // Ensure onProfileExists callback is called
         verify(signInRepository)
             .signInWithEmail(eq(email), eq(password)) // Ensure repository was called
   }
```
Comment (GhaliBerbich) [inline app/src/test/java/com/android/sample/model/auth/SignInViewModelTests.kt:252]: LGTM, changes made to match the new implementation of the mocked objects

File: app/src/main/java/com/android/sample/model/auth/SignInViewModel.kt [hunk 4]
```diff
@@ -68,6 +67,29 @@ constructor(
...
+            } else {
+              onProfileMissing()
+            }
+          },
+          {
+            Log.e("SignInViewModel", "Error checking user profile", it)
+            onFailure("Error checking user profile")
+          })
+    } ?: onFailure("UID is null")
+  }
```
Comment (GhaliBerbich) [inline app/src/main/java/com/android/sample/model/auth/SignInViewModel.kt:91]: Good job on oputting this function at its right place in the signInViewModel

File: app/src/main/java/com/android/sample/model/auth/SignInViewModel.kt [hunk 5]
```diff
@@ -19,47 +19,46 @@ open class SignInViewModel
...
       try {
         val authResult = signInRepository.signInWithEmail(email, password)
-        signInRepository.checkUserProfile(
-            authResult.user?.uid, navigationActions, onAuthSuccess, onAuthError)
+        checkUserProfile(authResult.user?.uid, onProfileExists, onProfileMissing, onFailure)
       } catch (e: Exception) {
-        onAuthError("Email login failed: ${e.message}")
+        Log.e("SignInViewModel", "Error signing in with email", e)
+        onFailure(e.message ?: "Unknown error")
       }
```
Comment (GhaliBerbich) [inline app/src/main/java/com/android/sample/model/auth/SignInViewModel.kt:40]: LGTM, adjusting to call the method checkUserProfile

File: app/src/main/java/com/android/sample/model/auth/SignInRepository.kt [hunk 6]
```diff
@@ -1,19 +1,11 @@
...
 
   suspend fun signInWithGoogle(idToken: String): AuthResult
 
-  fun checkUserProfile(
-      uid: String?,
-      navigationActions: NavigationActions,
-      onAuthSuccess: () -> Unit,
-      onAuthError: (String) -> Unit
-  )
-
```
Comment (GhaliBerbich) [inline app/src/main/java/com/android/sample/model/auth/SignInRepository.kt:17]: Yes, a function implemented in the viewmodel should not appear in the repository interface

File: app/src/main/java/com/android/sample/model/auth/SignInRepositoryFirebase.kt [hunk 7]
```diff
@@ -23,35 +20,6 @@ constructor(private val auth: FirebaseAuth, private val profilesRepository: Prof
...
-            Log.d("SignInRepository", "User profile found for user ID: $uid")
-            navigationActions.navigateTo(Screen.OVERVIEW)
-          }
-          onAuthSuccess()
-        },
-        onFailure = {
-          onAuthError("Error checking user profile!")
-          Log.e("SignInRepository", "Error checking user profile: ${it.message}")
-        })
-  }
```
Comment (GhaliBerbich) [inline app/src/main/java/com/android/sample/model/auth/SignInRepositoryFirebase.kt:53]: LGTM, deleting the function that you transferred to the viewmodel

**Review by GhaliBerbich — CHANGES_REQUESTED** [review]
The navigation logic has been moved out of the repository and into the screen, also the function checkUserProfile has been moved to the view model which is a positive step toward following the MVVM pattern. In the mockRepositoryFile, there is a TODO, if it is yours to complete, do it, if not indicate who should do it.

**Comment by aminebenji04** [comment]
> In the mockRepositoryFile, there is a TODO, if it is yours to complete, do it, if not indicate who should do it.

This is just a copy of the one already in androidtests which has this method lacking which is not needed for the moment in testing

**Review by GhaliBerbich — APPROVED** [review]

**Comment by GhaliBerbich** [comment]
> > In the mockRepositoryFile, there is a TODO, if it is yours to complete, do it, if not indicate who should do it.
> 
> This is just a copy of the one already in androidtests which has this method lacking which is not needed for the moment in testing
>>Ok I understand

**Pull request merged**